File: recover-binary-search-tree_iterative.py
# Time Complexity : O(N) whenre N is number of Nodes in the tree
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def recoverTree(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        first = None
        second = None
        flag  = False
        prev = None
        stack = []
        
        while root != None or len(stack) != 0:
            while root != None:
                stack.append(root)
                root = root.left
        
            root = stack.pop()
            if prev != None and prev.val >= root.val:
                if flag:
                    second = root
                else:
                    flag = True
                    first = prev
                    second = root

            prev = root
        
            root = root.right
        
        
        first.val , second.val = second.val,  first.val
        
        return root

File: Recover_Binary_Search_Tree_iterative.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// TC : O(n)
// SC : O(h)
// Method - 1 iteration
class Solution {
    public void recoverTree(TreeNode root) {
        if(root == null) return;
        
        Stack<TreeNode> st = new Stack<>();
        TreeNode first = null, second = null;
        TreeNode prev = null;
        while(root != null || !st.isEmpty()) {
            
            while(root != null) {
                st.push(root);
                root = root.left;
            }
            root = st.pop();
            if(prev != null && prev.val >= root.val) {
                if(first == null ) {
                     first = prev;
                     second = root;
                }
                    second = root;
            }
            prev = root;
            root = root.right;
        }
        int temp = first.val;
        first.val = second.val;
        second.val = temp;
    }
}
